File: frontend/src/shared/types/billing.ts
export type BillingInterval = "monthly" | "yearly";
export type SubscriptionStatus = "trialing" | "active" | "past_due" | "canceled" | "expired";
export type BillingTransactionStatus = "pending" | "succeeded" | "failed" | "refunded" | "canceled";
export type BillingTransactionType = "subscription" | "token_purchase" | "adjustment" | "refund";
export type TokenWalletStatus = "active" | "frozen" | "closed";
export type TokenLedgerEntryType = "credit" | "debit" | "reserve" | "release" | "expire" | "adjustment";
export type TokenAllocationType = "monthly_free" | "purchased" | "manual";

export type FeatureCatalogItem = {
  id: number;
  code: string;
  name: string;
  description: string | null;
  value_type: string;
  unit_label: string | null;
  is_active: boolean;
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type PlanFeature = {
  feature: FeatureCatalogItem;
  included_value: string;
  overage_price_usd: string | null;
  config_json: Record<string, unknown>;
};

export type BillingPlan = {
  id: number;
  code: string;
  name: string;
  description: string | null;
  billing_interval: BillingInterval;
  setup_fee_usd: string;
  price_usd: string;
  monthly_token_credit: number;
  is_active: boolean;
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  features: PlanFeature[];
};

export type BillingPlanList = {
  items: BillingPlan[];
  total: number;
};

export type FeatureCatalogList = {
  items: FeatureCatalogItem[];
  total: number;
};

export type AccountSubscription = {
  id: number;
  account_id: number;
  billing_plan_id: number;
  status: SubscriptionStatus;
  starts_at: string;
  ends_at: string | null;
  renews_at: string | null;
  canceled_at: string | null;
  external_subscription_id: string | null;
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  plan: BillingPlan;
};

export type TokenPackage = {
  id: number;
  code: string;
  name: string;
  description: string | null;
  token_amount: number;
  bonus_tokens: number;
  price_usd: string;
  currency: string;
  is_active: boolean;
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type TokenPackageList = {
  items: TokenPackage[];
  total: number;
};

export type BillingTransaction = {
  id: number;
  account_id: number;
  account_subscription_id: number | null;
  token_purchase_package_id: number | null;
  transaction_type: BillingTransactionType;
  status: BillingTransactionStatus;
  provider_name: string | null;
  external_reference: string | null;
  currency: string;
  amount_usd: string;
  tax_usd: string;
  total_amount_usd: string;
  occurred_at: string;
  paid_at: string | null;
  failed_at: string | null;
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type BillingTransactionList = {
  items: BillingTransaction[];
  total: number;
};

export type TokenLedgerEntry = {
  id: number;
  token_wallet_id: number;
  account_id: number;
  account_subscription_id: number | null;
  billing_transaction_id: number | null;
  entry_type: TokenLedgerEntryType;
  source_type: string;
  allocation_type: TokenAllocationType | null;
  delta_tokens: number;
  balance_before: number;
  balance_after: number;
  remaining_tokens: number | null;
  expires_at: string | null;
  expired_at: string | null;
  is_expired: boolean;
  unit_price_usd: string | null;
  total_price_usd: string | null;
  reference_type: string | null;
  reference_id: string | null;
  notes: string | null;
  occurred_at: string;
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type TokenLedgerList = {
  items: TokenLedgerEntry[];
  total: number;
};

export type TokenWallet = {
  id: number;
  account_id: number;
  status: TokenWalletStatus;
  balance_tokens: number;
  reserved_tokens: number;
  lifetime_credited_tokens: number;
  lifetime_debited_tokens: number;
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type TokenBalanceBreakdown = {
  available_monthly_free_tokens: number;
  available_purchased_tokens: number;
  available_manual_tokens: number;
  total_available_tokens: number;
  expiring_next_tokens: number;
};

export type WalletBalance = {
  wallet: TokenWallet;
  breakdown: TokenBalanceBreakdown;
};

File: backend/app/schemas/billing.py
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import (
    BillingInterval,
    BillingTransactionStatus,
    BillingTransactionType,
    SubscriptionStatus,
    TokenAllocationType,
    TokenLedgerEntryType,
    TokenLedgerSourceType,
    TokenWalletStatus,
)
from app.schemas.common import ORMModel


class BillingPlanResponse(ORMModel):
    id: int
    code: str
    name: str
    description: str | None
    billing_interval: BillingInterval
    setup_fee_usd: Decimal
    price_usd: Decimal
    monthly_token_credit: int
    is_active: bool
    metadata_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class FeatureCatalogResponse(ORMModel):
    id: int
    code: str
    name: str
    description: str | None
    value_type: str
    unit_label: str | None
    is_active: bool
    metadata_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class PlanFeatureResponse(BaseModel):
    feature: FeatureCatalogResponse
    included_value: Decimal
    overage_price_usd: Decimal | None = None
    config_json: dict[str, Any] = Field(default_factory=dict)


class BillingPlanDetailResponse(BillingPlanResponse):
    features: list[PlanFeatureResponse] = Field(default_factory=list)


class AccountSubscriptionResponse(ORMModel):
    id: int
    account_id: int
    billing_plan_id: int
    status: SubscriptionStatus
    starts_at: datetime
    ends_at: datetime | None
    renews_at: datetime | None
    canceled_at: datetime | None
    external_subscription_id: str | None
    metadata_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime
    plan: BillingPlanDetailResponse


class TokenWalletResponse(ORMModel):
    id: int
    account_id: int
    status: TokenWalletStatus
    balance_tokens: int
    reserved_tokens: int
    lifetime_credited_tokens: int
    lifetime_debited_tokens: int
    metadata_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class TokenBalanceBreakdown(BaseModel):
    available_monthly_free_tokens: int
    available_purchased_tokens: int
    available_manual_tokens: int
    total_available_tokens: int
    expiring_next_tokens: int


class WalletBalanceResponse(BaseModel):
    wallet: TokenWalletResponse
    breakdown: TokenBalanceBreakdown


class BillingTransactionResponse(ORMModel):
    id: int
    account_id: int
    account_subscription_id: int | None
    token_purchase_package_id: int | None
    transaction_type: BillingTransactionType
    status: BillingTransactionStatus
    provider_name: str | None
    external_reference: str | None
    currency: str
    amount_usd: Decimal
    tax_usd: Decimal
    total_amount_usd: Decimal
    occurred_at: datetime
    paid_at: datetime | None
    failed_at: datetime | None
    metadata_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class TokenLedgerResponse(ORMModel):
    id: int
    token_wallet_id: int
    account_id: int
    account_subscription_id: int | None
    billing_transaction_id: int | None
    entry_type: TokenLedgerEntryType
    source_type: TokenLedgerSourceType
    allocation_type: TokenAllocationType | None
    delta_tokens: int
    balance_before: int
    balance_after: int
    remaining_tokens: int | None
    expires_at: datetime | None
    expired_at: datetime | None
    is_expired: bool
    unit_price_usd: Decimal | None
    total_price_usd: Decimal | None
    reference_type: str | None
    reference_id: str | None
    notes: str | None
    occurred_at: datetime
    metadata_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class BillingPlanListResponse(BaseModel):
    items: list[BillingPlanDetailResponse]
    total: int


class FeatureCatalogListResponse(BaseModel):
    items: list[FeatureCatalogResponse]
    total: int


class TokenPackageResponse(ORMModel):
    id: int
    code: str
    name: str
    description: str | None
    token_amount: int
    bonus_tokens: int
    price_usd: Decimal
    currency: str
    is_active: bool
    metadata_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class TokenPackageListResponse(BaseModel):
    items: list[TokenPackageResponse]
    total: int


class BillingTransactionListResponse(BaseModel):
    items: list[BillingTransactionResponse]
    total: int


class TokenLedgerListResponse(BaseModel):
    items: list[TokenLedgerResponse]
    total: int


class SubscriptionCreateRequest(BaseModel):
    billing_plan_code: str = Field(..., min_length=2, max_length=64)
    status: SubscriptionStatus = SubscriptionStatus.ACTIVE
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    renews_at: datetime | None = None
    external_subscription_id: str | None = Field(default=None, max_length=255)
    metadata_json: dict[str, Any] = Field(default_factory=dict)


class TokenPackageCreateRequest(BaseModel):
    code: str = Field(..., min_length=2, max_length=100)
    name: str = Field(..., min_length=2, max_length=255)
    description: str | None = Field(default=None, max_length=500)
    token_amount: int = Field(..., ge=1)
    bonus_tokens: int = Field(default=0, ge=0)
    price_usd: Decimal = Field(..., ge=0)
    currency: str = Field(default="USD", min_length=3, max_length=8)
    metadata_json: dict[str, Any] = Field(default_factory=dict)


class TokenPackagePurchaseRequest(BaseModel):
    package_code: str = Field(..., min_length=2, max_length=100)
    metadata_json: dict[str, Any] = Field(default_factory=dict)


class ManualTokenAdjustmentRequest(BaseModel):
    delta_tokens: int
    notes: str | None = Field(default=None, max_length=500)
    metadata_json: dict[str, Any] = Field(default_factory=dict)
